shared/params.go

```diff
@@ -97,21 +181,54 @@ func ParseBrowsersParam(r *http.Request) (browsers []string, err error) {
 	return browsers, nil
 }
 
-// GetBrowsersForRequest parses the 'browsers' param, returning the sorted list of browsers
-// to include, or a default list.
-func GetBrowsersForRequest(r *http.Request) (browsers []string, err error) {
-	if browsers, err = ParseBrowsersParam(r); err != nil {
-		return browsers, err
+// ParseProductsParam returns a sorted list of product params for the request.
+// It parses the 'products' parameter, split on commas, and also checks for the (repeatable)
+// 'product' params.
+func ParseProductsParam(r *http.Request) (products []Product, err error) {
+	productParams := ParseRepeatedParam(r, "product", "products")
+	if productParams == nil {
+		return nil, nil
 	}
-	if browsers == nil || len(browsers) == 0 {
-		if browsers, err = GetBrowserNames(); err != nil {
-			return browsers, err
+	for p := range productParams.Iter() {
+		product, err := ParseProduct(p.(string))
+		if err != nil {
+			return nil, err
 		}
+		products = append(products, product)
 	}
+	return products, nil
+}
+
+// GetProductsForRequest parses the 'products' (and legacy 'browsers') params, returning
+// the sorted list of products to include, or a default list.
+func GetProductsForRequest(r *http.Request) (products []Product, err error) {
+	if products, err = ParseProductsParam(r); err != nil {
+		return nil, err
+	}
+	// Handle legacy browser param.
+	browserParams, err := ParseBrowsersParam(r)
+	if err != nil {
+		return nil, err
+	}
+	for _, browser := range browserParams {
+		products = append(products, Product{
+			BrowserName: browser,
+		})
+	}
+
+	browserNames, err := GetBrowserNames()
+	// Fall back to default browser set.
+	if products == nil && browserParams == nil {
```

Comment:
It doesn't look like `browserParams == nil` is possible at this point, is there a test that hits this code path? `products == nil` does look possible, so I'm guessing this should  check if `products` is nil or an empty list?

Reply:
In Golang, for slices, nil == empty.. they're reference types but you can always check `len([]interface{}(nil)) == 0`
https://play.golang.org/p/aLjLlOEBw8d

Note that there's 2 sets here - browser _params_ and browser _names_. Params can be empty, and so can products.

Reply:
OK, this is just more of Go that I don't know, empty slice is nil.